core/utils/codel.h

```diff
@@ -0,0 +1,267 @@
+  }
+
+  // Returns the capacity of the queue which one more than the total available
+  // slots in the ring buffer
+  size_t Capacity() { return queue_->Capacity(); }
+
+  // Returns true if empty
+  bool Empty() { return queue_->Empty(); }
+
+  // Returns true if full
+  bool Full() { return queue_->Full(); }
+
+  // Returns the number of objects in the queue
+  size_t Size() { return queue_->Size(); }
+
+ private:
+  // Calls the drop_func on the ptr and deallocates the Wrapper
+  // Takes ownership of a wrapper that contains the ptr to drop
+  void Drop(Wrapper* w) {
+    if (w != nullptr) {
+      if (w->ptr != nullptr) {
+        drop_func_(w->ptr);
+      }
+      delete w;
+    }
+  }
+
+  // Takes the relative time to determine the next time to drop.
+  // returns the next time to drop a packet.
+  uint64_t NextDrop(uint64_t cur_time) {
+    return cur_time + window_ * pow(drop_count_, -.5);
+  }
+
+  // Gets the next object from the queue and determines based on current state,
+  // whether to mark in the wrapper that the object should be dropped. Returns
+  // a wrapper with the next object from the queue if queue is not empty.
+  // Otherwise, returns nullptr.
+  Wrapper* RingDequeue() {
+    Wrapper* w = queue_->Pop();
+    if (!w) {
+      return nullptr;
+    }
+
+    uint64_t now = MicroSecondTime();
+    uint64_t delay_time = now - w->timestamp;
+
+    // determine whether object should be dropped or to change state
+    if (delay_time < delay_target_) {
+      time_above_target_ = 0;
+    } else {
+      if (time_above_target_ == 0) {
+        time_above_target_ = now + window_;
+      } else if (now >= time_above_target_) {
+        w->drop = 1;
+      }
+    }
+    return w;
+  }
+
+  // Called while Codel is in drop state to determine whether to drop the current
+  // entries and dequeue the next entry. Will continue to drop entries until the 
+  // next time till drop is greater than the current time. Takes ownership of the
+  // wrapper pointer where the current entry is located.
+  Wrapper* DropDequeue(Wrapper* w) {
+    uint64_t now = MicroSecondTime();
+    if (!w->drop) {
+      dropping_ = 0;
+    } else if (now >= time_till_drop_) {
+      while (now >= time_till_drop_ && dropping_) {
+        Drop(w);
+        drop_count_++;
+        w = RingDequeue();
+
+        if (!w || !w->drop) {
+          dropping_ = 0;
+        } else {
+          time_till_drop_ = NextDrop(time_till_drop_);
+        }
+      }
+    }
+    return w;
+  }
+
+  // Returns the current time in microseconds.
+  uint64_t MicroSecondTime() {
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+    return tv.tv_sec * 1e6 + tv.tv_usec;
+  }
+
+  uint64_t delay_target_;  // the delay that codel will adjust for
+  uint64_t window_;        // minimum time before changing state
+
+  // the time at which codel will change state to above target(0 if below)
+  uint64_t time_above_target_;
+  uint64_t time_till_drop_;  // the next time codel will drop
```

Comment:
Rename this member variable -- if I understand it right, it's not the time *until* drop, but rather the time *for* the next drop (i.e. absolute rather than relative).